ngrok Vaults + Secrets  docs (#1371)
ngrok/ngrok-docs: docs/traffic-policy/macros/security.mdx

        config:
          realm: sample-realm
          credentials:
            - \${secret('example-vault','example-basic-auth')}
          enforce: true
```

## Using secrets in you Kubernetes deployment

Vaults and secrets can be used in traffic policies in conjunction with Kubernetes Operator. Vaults and Secrets must be created using the ngrok API (either through REST or via the CLI).

You can use your secrets with your Kubernetes deployment via Traffic Policy. See [the Kubernetes Operator quickstart](/getting-started/kubernetes/ingress) to learn more.

## Macros

### `secret(string, string) -> string`

Takes the vault name as the first argument and the secret name as the second argument. Returns the secret value.

#### Example

```yaml
secret("vault-name", "secret-name")
```

## Supported Traffic Policy Actions

- [`set-vars`](/traffic-policy/actions/set-vars/)
  - CEL is supported in the following fields:
    - `vars`
- [`verify-webhook`](/traffic-policy/actions/verify-webhook/)
  - CEL is supported in the following fields:
    - `secret`
- [`basic-auth`](/traffic-policy/actions/basic-auth/)
  - CEL is supported in the following fields:
    - `credentials`

> - `[jwt-validation`](/traffic-policy/actions/jwt-validation/)

Author: Works for me (you don't have to put `/docs/` in front of links in the docs repo)

> - `[jwt-validation`](/traffic-policy/actions/jwt-validation/)

Author: Ohhh, I see, good catch!

---

> title: Security Macros

Reviewer: ```suggestion
title: Security Macros (Early Access)
```

description: Security Macros allow you to access sensitive information directly in your Traffic Policies. 
---

Security Macros allow you to access sensitive information directly in your Traffic Policies. Your ngrok account has a Vault that can store Secrets. Any secrets that you add to your vault will be available across your account on all traffic policies. Updates to these secrets will be reflected across all traffic policies automatically.

## How secrets are secured

- Secrets are protected at rest using industry standard `AES-256` encryption
- ngrok's REST API does not return secrets as part of any of its response payloads
- REST API traffic is encrypted in-transit using `HTTP/S` and `TLS 1.2+`

## Using secrets in you Kubernetes deployment

> TBD

Reviewer: @S3Prototype OK, synced with Kate.

The primary thrust of what we need to document here is that Vaults & Secrets need to be created through the REST API or CLI + API, that they cannot be created using k8s directly.

Kate put together a small tutorial [here](https://linear.app/ngrok/document/testing-with-kubernetes-operator-d33f7d98c581), showing usage. This could make a great guide. I think for the purposes of this doc at this point in time, however, we can get by simply by documenting something akin to the following:

```
Vaults and secrets can be used in traffic policies in conjunction with Kubernetes Operator. Vaults and Secrets must be created using the ngrok API (either through REST or via the CLI).
```

Optionally, we show a snippet - which would map to steps \#1 & \#2 in Kate's doc.

> TBD

Author: Thanks for the info!

> TBD

Reviewer: I created a workitem for a How-To Guide on this. I think this can come post-GA as long as we cover the above ^^ points in this doc for Early Access / GA

Workitem:
https://linear.app/ngrok/issue/EPD-1147/create-a-how-to-guide-for-k8s-vaultssecrets-usage

## Macros

### `secret(string, string) -> string`

Takes the vault name as the first argument and the secret name as the second argument. Returns the secret value.

> 

Reviewer: We should also mention which actions are supported while in Early Access
- `set-vars`
  - CEL is supported in the following fields:
    - `vars`
- `verify-webhook`
    - CEL is supported in the following fields:
      - `secret`
- `basic-auth`
   - CEL is supported in the following fields:
     - `credentials`
- `jwt-validation`
  - CEL is supported in the following fields:
    - `issuer.allow_list[*].value`
    - `http.tokens[*].name`
    - `http.tokens[*].prefix`
    - `jws.keys[*].sources[*].additional_jkus`

#### Example

```yaml
secret("vault-name", "secret-name")
```

## Pricing and limits

Each vault and secret is free. Price per action is TBD.

| Plan     | Vault Limit     | Secret Limit (across all vaults) | Notes                                                                             |
| -------- | --------------- | -------------------------------- | --------------------------------------------------------------------------------- |
| Free     | 5 (hard–capped) | 5 (hard–capped)                  | All values are fixed. If you need more vaults/secrets, move to Personal or Paygo. |

> | Pro      | 10              | 100                              | All values are fixed. If you need more vaults/secrets, move to Personal or Paygo. |

Reviewer: ```suggestion
| Pro      | 5              | 25                              | All values are fixed. If you need more vaults/secrets, move to Paygo. |
```

> | Business | 100             | 1000                             | Secret limit is on a slider.                                                      |

Reviewer: ```suggestion
| Business | 5 (default limit)             | 500 (default limit)        | Contact us if you need more vaults/secrets than default limits |
```
